### PR TITLE
Comonad laws

### PR DESCRIPTION
**Pull request by pakoito**
This diff adds laws for Comonad, along with tests for all instances.

Turns out Cofree was correctly implemented after all. Huh.

### CODE REVIEW

File: kategory-test/src/main/kotlin/kategory/laws/TraverseLaws.kt [hunk 1]
```diff
@@ -17,29 +17,39 @@ class TIF {
...
+    // FIXME(paco): this implementation will crash the inliner
+    /*
+    inline fun <reified F> laws(TF: Traverse<F> = traverse<F>(), AF: Applicative<F> = applicative<F>(), EQ: Eq<HK<F, Int>>): List<Law> =
+        FoldableLaws.laws(TF, { AF.pure(it) }, Eq.any()) + FunctorLaws.laws(AF, EQ) + listOf(
+                Law("Traverse Laws: Identity", { identityTraverse(TF, AF, { AF.pure(it) }, EQ) }),
+                Law("Traverse Laws: Sequential composition", { sequentialComposition(TF, { AF.pure(it) }, EQ) }),
+                Law("Traverse Laws: Parallel composition", { parallelComposition(TF, { AF.pure(it) }, EQ) }),
+                Law("Traverse Laws: FoldMap derived", { foldMapDerived(TF, { AF.pure(it) }) })
+        )
+    */
```
Comment (JorgeCastilloPrz) [inline kategory-test/src/main/kotlin/kategory/laws/TraverseLaws.kt]: Shouldn't this comment get removed ?

Comment (pakoito) [inline kategory-test/src/main/kotlin/kategory/laws/TraverseLaws.kt]: No, it has to be re-enabled after the issue in the fixme is fixed

Comment (JorgeCastilloPrz) [inline kategory-test/src/main/kotlin/kategory/laws/TraverseLaws.kt]: So, it's gonna be fixed prior to merging this or in a different PR ?

Comment (pakoito) [inline kategory-test/src/main/kotlin/kategory/laws/TraverseLaws.kt]: It's an issue on the Kotlin compiler, I may beed to add a reference to it: KT-18660

**Pull request merged**